str.c

```diff
@@ -723,6 +728,98 @@ str_replace_unprintable(struct mystr* p_str, char new_char)
   }
 }
 
+void
+str_replace_unprintable_with_hex(struct mystr* p_str)
+{
+  unsigned int ups_size = sizeof(unsigned int) * (p_str->len);
+  if (ups_size < p_str->len)
+  {
+    bug("string is to long");
+  }
+  unsigned int* ups = vsf_sysutil_malloc(ups_size);
+  unsigned int up_count = 0;
+  for (unsigned int i=0; i < p_str->len; i++)
+  {
+    if (!vsf_sysutil_isprint(p_str->p_buf[i]))
+    {
+      ups[up_count++] = i;
+    }
+  }
+  str_replace_positions_with_hex(p_str, ups, up_count);
+  vsf_sysutil_free(ups);
+}
+
+void str_replace_unprintable_with_hex_wc(struct mystr* p_str)
+{
+  unsigned int ups_size = sizeof(unsigned int) * (p_str->len);
+  if (ups_size < p_str->len)
+  {
+    bug("string is to long");
+  }
+  unsigned int* ups = vsf_sysutil_malloc(ups_size);
+  unsigned int up_count = 0;
+
+  size_t current = 0;
+  wchar_t pwc;
+  mbstate_t ps;
+  memset(&ps, 0, sizeof(ps));
+  ssize_t len = 0;
+  while ((len = mbrtowc(&pwc, p_str->p_buf, p_str->len - current, &ps)) > 0)
+  {
+    if (!iswprint(pwc))
+    {
+      for (unsigned int i = 0; i < len; i++)
+      {
+        ups[up_count++] = current++;
+      }
+    }
+    else
+    {
+      current += len;
+    }
+  }
+  if (len < 0)
+  {
+    while (current < p_str->len)
+    {
+      ups[up_count++] = current++;
+    }
+  }
+  str_replace_positions_with_hex(p_str, ups, up_count);
+  vsf_sysutil_free(ups);
+}
+
+void
+str_replace_positions_with_hex(struct mystr* p_str, const unsigned int* poss, const unsigned int pos_count)
+{
+  if (pos_count == 0)
+    return;
+
+  struct mystr tmp_str = INIT_MYSTR;
+  str_reserve(&tmp_str, p_str->len + 3 * pos_count);
+  unsigned int current = 0;
+
+  for (unsigned int i=0; i < pos_count; i++)
+  {
+    unsigned int pos = poss[i];
+
+    if (current < pos)
+      private_str_append_memchunk(&tmp_str, p_str->p_buf + current, pos - current);
+
+    char hex_buf[5];
+    memset(hex_buf, 0, sizeof(hex_buf));
+    sprintf(hex_buf, "\\x%02X", (unsigned char) p_str->p_buf[pos]);
```

Comment:
%02X format is too short for wide characters. Does it handle bytes as single byte encoding or multibyte? It seems no clear distinction between both modes. Not sure why iswprint() is used on different place, it seems inconsistent.

Reply:
1. At this point, individual bytes are converted to hex representation. ``` p_str->p_buf[pos]``` has type char. So ```%02X``` will be enough for value with a size of one byte.
2. The distinction between functions is that one treats the input string as a plain ASCII string (one byte = one character) using ```vsf_sysutil_isprint``` to determine if the byte is printable and the second one treats it as a wide-character string (_wc one) using ```iswprint``` to determine if an array of bytes is printable wide-character. User can choose which one will be used with ```wc_logs_enable``` configuration parameter.

Reply:
Please check my comment below. It should provide information what is the difference between the two functions and why ```iswprint``` is used in different place.

Reply:
I would prefer "\\u%02X", (wchar_t) decoded_wchar. It would print binary encoded unprintable characters, but it decoded it successfully when checking iswprint on it. But I guess main reason to print hex was usually just unicode characters, so people should not see them often.